[Feature] JavaScript Addons Support

infinyte7 opened the pull request:
## Pull Request template

## Purpose / Description
Using AnkiDroid JS API a deck can be extended to implement features like progress bar or custom card layouts. But it requires user to manually add the ```html/css/js``` to card templates. This is implementation for a way to inject code to all deck templates when addons get activated.

## Fixes
Fixes #7959 

## Approach
1. Download addons from https://www.npmjs.com/search?q=keywords:ankidroid-js-addon inside the app
2. In app there are implementation to check for valid addons for AnkiDroid
It will check following string in latest package.json
```
{
    "name": "ankidroid-js-addon-progress-bar",
    "version": "1.0.0",
    "author": "https://github.com/infinyte7",
    "homepage": "https://github.com/infinyte7/Anki-Custom-Card-Layout",
    "ankidroid_js_api": "0.0.1",
    "addon_type": "reviewer",
    "keywords": ["ankidroid-js-addon"]
}
```
There isValidPackageJson boolean function will check for above values if present then the addons is valid.

3. The ```enabled``` addons' ```index.js``` content wrapped in ```<script>``` tag will be added to ```fillFlashCard()``` in ```AbstractFlashcardViewer.java```

I have also documented it here https://github.com/infinyte7/ankidroid-js-addon

There are two types of addon ```reviewer``` and ```note_editor```. In this PR, reviewer addon is implemented.

<img src="https://user-images.githubusercontent.com/12841290/103460959-aaeab700-4d55-11eb-9226-2bec784b9871.gif" height=400/>

## How Has This Been Tested?
Tested on emulator 

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)


## Learning (optional, can help others)
https://developer.android.com/guide/topics/ui/layout/recyclerview

## Checklist
_Please, go through these checks before submitting the PR._

- [ x ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ x ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ x ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ x ] You have commented your code, particularly in hard-to-understand areas
- [ x ] You have performed a self-review of your own code
- [ x ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

david-allison-1 commented on AnkiDroid/build.gradle:
```diff
@@ -298,4 +298,7 @@ dependencies {
     androidTestImplementation 'androidx.test:rules:1.3.0'
+
+    // for untar .tgz file downloaded from npmjs for AnkiDroid js addon support
+    implementation 'org.rauschig:jarchivelib:1.1.0'
```
nit: move to where other `implementation` directives are

david-allison-1 commented on AnkiDroid/src/main/AndroidManifest.xml:
```diff
@@ -322,6 +322,18 @@
+            >
+            <!-- The meta-data element is needed for versions lower than 4.1 -->
+            <meta-data
+                android:name="android.support.PARENT_ACTIVITY"
```
Is this still needed?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2402,6 +2416,83 @@ private static long getDeckIdForCard(final Card card) {
+        Map<String,?> keys = preferences.getAll();
+
+        for(Map.Entry<String,?> entry : keys.entrySet()){
+            Timber.d("map values %s",entry.getKey() + ": " + entry.getValue().toString());
```
```suggestion
            Timber.d("map values %s: %s", entry.getKey(), entry.getValue());
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2402,6 +2416,83 @@ private static long getDeckIdForCard(final Card card) {
+                    }
+                }
+                catch (ArrayIndexOutOfBoundsException | IOException e) {
+                    Timber.e(e, "AbstractFlashcardViewer::IOException");
```
```suggestion
                    Timber.w(e, "AbstractFlashcardViewer::IOException");
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2402,6 +2416,83 @@ private static long getDeckIdForCard(final Card card) {
+     */
+    public String getEnabledAddonsContent() {
+        StringBuilder content = new StringBuilder();
+        String mainJsFile = null;
```
I think this would be better in a separate class

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2402,6 +2416,83 @@ private static long getDeckIdForCard(final Card card) {
+            Timber.d("map values %s",entry.getKey() + ": " + entry.getValue().toString());
+
+            // get enabled status of addons in SharedPreferences with key containing 'addon'
+            if (entry.getKey().contains("reviewer_addon") && entry.getValue().toString().equals("enabled")) {
```
Invert this `if`, and `continue;` if it's false

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2402,6 +2416,83 @@ private static long getDeckIdForCard(final Card card) {
+                    String addonDirName = String.valueOf(entry.getKey().split(":")[1]);
+                    File finalAddonPath = new File(addonsHomeDir, addonDirName);
+
+                    if (finalAddonPath.exists()) {
```
Again: invert the if to reduce nesting

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2402,6 +2416,83 @@ private static long getDeckIdForCard(final Card card) {
+        String mainJsFile = null;
+
+        SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(this);
+        String currentAnkiDroidDirectory = CollectionHelper.getCurrentAnkiDroidDirectory(this);
```
This method is too long, and too nested, I've added some small suggestions, but some things could be better encapsulated in a class.

I'll follow up from this in another review

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2410,7 +2501,13 @@ public void fillFlashcard() {
+        if (jsAddonContent == null) {
+            jsAddonContent = "";
+        }
+        final String cardContent = mCardContent + jsAddonContent;
```
efficiency: I think we'll be able to move this to improve the performance hit

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,50 @@
+package com.ichi2.anki;
+
+public class AddonModel {
+    private String name;
```
Naming conventions
```java
    private String sName;
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,102 @@
+            } else {
+                editor.putString(addonModel.getType() + "_addon:" + addonModel.getName(), "disabled");
+                editor.apply();
+                UIUtils.showThemedToast(context, context.getString(R.string.disabled) + " " + addonModel.getName(), true);
```
Don't concatenate strings, this implies an order, ask for another translation

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,102 @@
+
+        // while binding viewholder if preferences w.r.t viewholder store true value or enabled status then
+        // turn on switch status else it is off by default
+        if (preferences.getString(addonModel.getType() + "_addon:" + addonModel.getName(), "disabled").equals("enabled")) {
```
extract: `"_addon:" + addonModel.getName()` into a variable - makes the code easier to read

david-allison-1 replied:
Is there a reason that you're using a string here instead of a boolean?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,102 @@
+            SharedPreferences.Editor editor = preferences.edit();
+            // store enabled/disabled status as boolean true/false value in SharedPreferences
+            if (holder.addonActivate.isChecked()) {
+                editor.putString(addonModel.getType() + "_addon:" + addonModel.getName(), "enabled");
+                editor.apply();
```
```suggestion
                editor.putString(addonModel.getType() + "_addon:" + addonModel.getName(), "enabled").apply();
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,102 @@
+        holder.addonActivate.setOnClickListener(v -> {
+            SharedPreferences.Editor editor = preferences.edit();
+            // store enabled/disabled status as boolean true/false value in SharedPreferences
+            if (holder.addonActivate.isChecked()) {
```
This method can be simplified further

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,102 @@
+
+    public class AddonsViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
+        TextView addonsTextView;
+        @SuppressLint("UseSwitchCompatOrMaterialCode")
```
What is this error? Why is it suppressed?

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/AddonsBrowserTest.java:
```diff
@@ -0,0 +1,7 @@
+package com.ichi2.anki;
+
+public class AddonsBrowserTest extends RobolectricTest {
```
This shouldn't need to exist, or it should, but should have tests

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,576 @@
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        Themes.setThemeLegacy(this);
```
Is this needed?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,576 @@
+        public void onReceive(Context context, Intent intent) {
+            String action = intent.getAction();
+
+            if (DownloadManager.ACTION_DOWNLOAD_COMPLETE.equals(action)) {
```
See if you can invert some of these if statements to reduce nesting

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,576 @@
+
+
+    public static void deleteDirectory(File dir) {
+        if (dir.isDirectory()) {
```
invert the if

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,576 @@
+            for (int i = 0; i < children.length; i++) {
+                File child = new File(dir, children[i]);
+
+                if (child.isDirectory()) {
+                    deleteDirectory(child);
+                    child.delete();
+                } else {
+                    child.delete();
+                }
```
```suggestion
                if (child.isDirectory()) {
                    deleteDirectory(child);
                }
                child.delete();
```

infinyte7 commented:
How it will works?
1. Users will navigate to Addons Browser Activity
2. There are four button in options menu
     - Reviewer Addons
     - Note editor Addons  (Separate PR)
     - Import Addons
     - Get More Addons
3. When user click `Import Addons`
It will ask to paste `npm i ankidroid-js-addons...` that is available on npm package on npmjs.com
4. When user click `Download` it will download the addons to `AnkiDroid/addons` folder
5. There are implementation for checking valid npm package
```
{
    "name": "ankidroid-js-addon-progress-bar",
    "version": "1.0.0",
    "author": "https://github.com/infinyte7",
    "homepage": "https://github.com/infinyte7/Anki-Custom-Card-Layout",
    "ankidroid_js_api": "0.0.1",
    "addon_type": "reviewer",
    "keywords": ["ankidroid-js-addon"]
}
```
If above found in package.json on npmjs registry then then it will download the package.json and get tarball url.
6. Here package with `.tgz` extension will be downloaded to `Downloads` folder and extracted then copied to `AnkiDroid/addons`  folder
7. The addons will be listed on the Addons Browser screen
8. Users have option to enable, update and delete addons.
9. When users enable the addons, the `index.js` file of that addons will be added to card during review. The addons have full access to AnkiDroid JS API.

**Demo of progress bar**
<img src="https://user-images.githubusercontent.com/12841290/113458708-c1340c00-9445-11eb-85bb-985aba56cb2a.gif" height=500></img>

david-allison-1 reviewed (commented):
This is a partial review, sorry, I'm tired and unable to go into depth right now. It's on me to re-review.

It looks good from a brief scan! And it'll be great to get this in. I think a few areas could do with refactoring for readability, which will make a future structural review easier.

My main thought from a design/maintenance perspective is that there should be more classes, in order to explain NPMs structure to a developer who is unfamiliar with how packages work, and what they contain, and insulating a developer from implementation details, showing at a high level what the algorithms aim to accomplish.

There's a lot of logic in a few of the methods, and it'd be better to break some of them up

infinyte7 commented:
Thanks, I will update it with suggested changes.

infinyte7 commented:
>My main thought from a design/maintenance perspective is that there should be more classes, in order to explain NPMs structure to a developer who is unfamiliar with how packages work, and what they contain, and insulating a developer from implementation details, showing at a high level what the algorithms aim to accomplish.

I am implementing this so all npm related function should be in one class.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -1,50 +1,50 @@
-    private String jsApiVersion;
-    private String homepage;
-    private String type;
+    private String sName;
```
m, not s: https://github.com/ankidroid/Anki-Android/wiki/Code-style
```java
    private String sName;
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
 
-                if (cursor.moveToFirst()) {
-                    int columnIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
+            Cursor cursor = downloadManager.query(query);
```
Wrap this in try with resources

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
 
-                    if (DownloadManager.STATUS_SUCCESSFUL == cursor.getInt(columnIndex)) {
-                        String addonName = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_TITLE));
+            if (cursor.moveToFirst()) {
```
invert if to reduce nesting

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
+                    hideProgressBar();
+                    downloadManager.remove(queueId);
+                    UIUtils.showThemedToast(context, getString(R.string.error_downloading_file_check_name), false);
+                }
```
is there another case here? what should happen? I feel there should be an `else`, not an `else if`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
-                        String tarballPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
-                                + "/" + Uri.parse(addonName).getPath() + ".tgz";
-                        File tarballFile = new File(tarballPath);
+                if (DownloadManager.STATUS_SUCCESSFUL == cursor.getInt(columnIndex)) {
```
Invert if and return

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
-                            hideProgressBar();
-                        }
+                    File addonsDir = new File(addonsHomeDir, addonName);
+                    String tarballPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
```
Deprecated in Android Q

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
-                        } else {
-                            hideProgressBar();
-                        }
+                    File addonsDir = new File(addonsHomeDir, addonName);
```
Security: if addon name contains `../`, can the user write to an arbitrary path?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -557,20 +560,103 @@ public void onAddonClick(int position) {
+            return;
+        }
+
+        String[] children = dir.list();
```
Likely better as `.listFiles()`?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -557,20 +560,103 @@ public void onAddonClick(int position) {
+                    content.append("<script>");
+                    content.append("\n");
+
+                    BufferedReader br = new BufferedReader(new FileReader(addonsContentFile));
```
Use `try with resources`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -557,20 +560,103 @@ public void onAddonClick(int position) {
+                    content.append("<script>");
+                    content.append("\n");
+
+                    BufferedReader br = new BufferedReader(new FileReader(addonsContentFile));
+                    String line;
+
+                    while ((line = br.readLine()) != null) {
+                        content.append(line);
+                        content.append('\n');
+                    }
+                    br.close();
```
Extract this to a method, check to see if we have a utility method which does this already

infinyte7 replied:
I tried it.
Utils.convertStreamToString(is) not used because in that method `content.append('\n')` not present so it make js code not to work
I used `Utils.convertStreamToString(is)` for reading json file and there are no problem to them.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -557,20 +560,103 @@ public void onAddonClick(int position) {
+
+                // read index.js file or file pointed by "main"
+                File addonsContentFile = new File(addonsPackageDir, mainJsFile);
+                if (addonsContentFile.exists()) {
```
invert the if

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/testutils/ActivityList.java:
```diff
@@ -91,8 +91,7 @@
                 get(CardTemplateEditor.class, ActivityList::intentForCardTemplateEditor),
-                get(CardTemplateBrowserAppearanceEditor.class, ActivityList::intentForCardTemplateBrowserAppearanceEditor),
-                get(AddonsBrowser.class)
+                get(CardTemplateBrowserAppearanceEditor.class, ActivityList::intentForCardTemplateBrowserAppearanceEditor)
```
I believe this is still necessary and shouldn't have been reverted

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -364,55 +368,53 @@ public void downloadAddonPackageFile(JSONObject jsonObject, String npmAddonName)
-                        }
+                    File addonsDir = new File(addonsHomeDir, addonName);
+                    String tarballPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
+                            + "/" + Uri.parse(addonName).getPath() + ".tgz";
```
`Uri.parse(addonName).getPath()` can fail, what happens?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+    private Dialog downloadDialog;
+
+    // Update if api get updated
+    private String AnkiDroidJsAPI = "0.0.1";
```
Shouldn't these refer to the same values as `AbstractFlashcardViewer`, maybe move to a constants class

infinyte7 replied:
I will create separate PR to add API constant from every .java files

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+
+    private SharedPreferences preferences;
+
+    private String[] addonTypes = {"reviewer", "note_editor"};
```
We only handle the reviewer here

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+
+        addonsMessageLayout = findViewById(R.id.addons_download_message_ll);
+        addonsMessageLayout.setOnClickListener(v -> {
+            openUrl(Uri.parse(getResources().getString(R.string.ankidroid_js_addon_npm_search_url)));
```
I think:

```java
            openUrl(Uri.parse(getString(R.string.ankidroid_js_addon_npm_search_url)));
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(this, DividerItemDecoration.VERTICAL);
+        addonsListRecyclerView.addItemDecoration(dividerItemDecoration);
+
+        downloadDialog = new Dialog(this);
```
This seems like a nonstandard way to handle a dialog

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+
+        // default list reviewer addons
+        getSupportActionBar().setTitle(getString(R.string.reviewer_addons));
+        listAddonsFromDir(addonTypes[0]);
```
I think

```suggestion
        listAddonsFromDir(listAddonType);
```

infinyte7 replied:
The list addon type is for 
* Reviewer 
* Note editor
So when user click reviewer it shows reviewer addons in recycler view else note editor addons

infinyte7 replied:
I have removed the note editor options from menu. Now only `reviewer`.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+public class AddonsBrowser extends NavigationDrawerActivity implements DeckDropDownAdapter.SubtitleListener, AddonsAdapter.OnAddonClickListener {
+
+    private RecyclerView addonsListRecyclerView;
+    private MenuItem mInstallAddon;
```
These variables can be local to the function

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+    private MenuItem mReviewerAddons;
+    private MenuItem mNoteEditorAddons;
+
+    private long queueId;
```
This appears unused

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+                JSONObject jsonObject = AddonsNpmUtility.packageJsonReader(addonsPackageJson);
+
+                AddonModel addonModel;
+                if (isValidAddonPackage(jsonObject)) {
```
If it's invalid, do we want to let the user know?

infinyte7 replied:
I tried to showtoast but didn't work I got error.
`
Can't toast on a thread that has not called Looper.prepare()
`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+            File directory = new File(String.valueOf(addonsHomeDir));
+            File[] files = directory.listFiles();
+            for (File file : files) {
+                Timber.d("Addons:%s", file.getName());
```
```suggestion
                Timber.d("Addons: %s", file.getName());
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+                JSONObject jsonObject = AddonsNpmUtility.packageJsonReader(addonsPackageJson);
+
+                AddonModel addonModel;
+                if (isValidAddonPackage(jsonObject)) {
+
+                    String addonName = jsonObject.optString("name", "");
+                    String addonVersion = jsonObject.optString("version", "");
+                    String addonDev = jsonObject.optString("author", "");
+                    String addonAnkiDroidAPI = jsonObject.optString("ankidroid_js_api", "");
+                    String addonHomepage = jsonObject.optString("homepage", "");
+                    String typeOfAddon = jsonObject.optString("addon_type", "");
+
+                    if (addonType.equals(typeOfAddon)) {
+                        addonModel = new AddonModel(addonName, addonVersion, addonDev, addonAnkiDroidAPI, addonHomepage, addonType);
+                        addonsNames.add(addonModel);
+                    }
+                }
```
I would consider extracting to `AddonModel.tryParse(jsonObject)`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+
+            addonsListRecyclerView.setAdapter(new AddonsAdapter(addonsNames, this));
+
+            if (addonsNames.isEmpty()) {
+                addonsMessageLayout.setVisibility(View.VISIBLE);
+            } else {
+                addonsMessageLayout.setVisibility(View.GONE);
+            }
```
I don't know what this does without context (what is `addonsMessageLayout`?)

Consider:
* Including a comment as to why this happens, 
* Extract to a named method where the method name gives context
* Renaming the variable to be more descriptive

```suggestion
            addonsMessageLayout.setVisibility(addonsNames.isEmpty() ? View.VISIBLE : View.GONE);
```

infinyte7 replied:
This is extension icon with message below on Addons Browser first screen when no addons available.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+                        deleteDirectory(dir);
+
+                        // remove enabled status
+                        SharedPreferences.Editor editor = preferences.edit();
+                        editor.remove(addonModel.getType() + "_addon:" + addonModel.getName());
+                        editor.apply();
```
It'd also be useful to extract this string `addonModel.getType() + "_addon:" + addonModel.getName()` into the AddonModel class

```suggestion
                        preferences.edit().remove(addonModel.getType() + "_addon:" + addonModel.getName()).apply();
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+                        editor.apply();
+
+                        addonsNames.remove(position);
+                        Objects.requireNonNull(addonsListRecyclerView.getAdapter()).notifyItemRemoved(position);
+                        addonsListRecyclerView.getAdapter().notifyItemRangeChanged(position, addonsNames.size());
+                        addonsListRecyclerView.getAdapter().notifyDataSetChanged();
```
Do all 3 need to be called?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,438 @@
+                    getString(R.string.dialog_cancel),
+                    (dialog, id) -> dialog.cancel());
+
+            AlertDialog deleteAlert = alertBuilder.create();
```
I think `alertBuilder` has a `.show()` method

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,235 @@
+            return false;
+        }
+
+        String addonName = jsonObject.optString("name", "");
+        String addonVersion = jsonObject.optString("version", "");
+        String addonDev = jsonObject.optString("author", "");
+        String addonAnkiDroidAPI = jsonObject.optString("ankidroid_js_api", "");
+        String addonHomepage = jsonObject.optString("homepage", "");
```
(I might have posted this elsewhere) - this, and a lot of this method's functionality would likely be better inside the class

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,235 @@
+            }
+            Timber.d("keywords %s", keywords.toString());
+        } catch (JSONException e) {
+            Timber.e(e.getLocalizedMessage());
```
```suggestion
            Timber.w(e, e.getLocalizedMessage());
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,235 @@
+
+                    File tarballFile = new File(addonsHomeDir, npmAddonName + ".tgz");
+
+                    is = response.body().byteStream();
+                    fos = new FileOutputStream(tarballFile);
+
+                    while ((len = is.read(buf)) != -1) {
+                        fos.write(buf, 0, len);
+                    }
+                    fos.flush();
```
We should have util methods which do this already if you search the codebase:

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,235 @@
+
+                    File tarballFile = new File(addonsHomeDir, npmAddonName + ".tgz");
+
+                    is = response.body().byteStream();
```
Try with resources

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,235 @@
+
+                    extractAndCopyAddonTgz(tarballFile.getPath(), npmAddonName, context);
+                } catch (IOException | NullPointerException e) {
+                    Timber.e(e.getLocalizedMessage());
```
```suggestion
                    Timber.w(e, e.getLocalizedMessage());
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,235 @@
+    public static JSONObject packageJsonReader(File addonsFiles) {
+        JSONObject jsonObject = null;
+        try {
+            FileReader fileReader = new FileReader(addonsFiles);
+            BufferedReader bufferedReader = new BufferedReader(fileReader);
+            StringBuilder stringBuilder = new StringBuilder();
+            String line = bufferedReader.readLine();
+            while (line != null) {
+                stringBuilder.append(line).append("\n");
+                line = bufferedReader.readLine();
+            }
+            bufferedReader.close();
+
```
Use shared utils, and try with resources

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/testutils/ActivityList.java:
```diff
@@ -20,6 +20,7 @@
 import android.content.Context;
 import android.content.Intent;
 
+import com.ichi2.anki.AddonsBrowser;
```
Necessary?

david-allison-1 reviewed (commented):
Much nicer - second round, mostly structuring again

infinyte7 commented:
Thanks again, I will update it.

david-allison-1 commented:
Just to note; the idea of inverting `if`s is to remove the horizontal nesting: this means that someone reading the code has less to think about, it's a mechanically simple change which makes the code a lot more readable I saw that one change was reverted, and I might have had a comment misunderstood.

As a general rule: prefer flatter functions

infinyte7 commented:
I will again make more simple implementation.

infinyte7 commented:
Implementation:
I have tried to called toast and progress bar method but it crashes AnkiDroid, so I have implemented static method for progress bar and showing toast. It is working.
Also cleaned and restructured code a little  bit.

It become faster and simpler than my original implementation. 
Thanks

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+
+        preferences = AnkiDroidApp.getSharedPrefs(this);
+
+        // default list reviewer addons
+        getSupportActionBar().setTitle(getString(R.string.reviewer_addons));
```
Might as well move this to where the other `getSupportActionBar()` methods are called

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+
+                // if string is:  npm i ankidroid-js-addon-progress-bar
+                if (npmAddonName.startsWith("npm i")) {
+                    npmAddonName = npmAddonName.replace("npm i", "");
```
Use a substring with `"npm i".length()`; `replace` will replace all instances within the string

infinyte7 replied:
I didn't understand. Please provide more info.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+                }
+
+                // if containing space
+                npmAddonName = npmAddonName.trim();
+                npmAddonName = npmAddonName.replaceAll("\u00A0", "");
```
TODO: Check that this is what you want

```suggestion
                npmAddonName = StringUtil.strip(npmAddonName);
```

infinyte7 replied:
It didn't worked

david-allison-1 replied:
Best to get a unit test in to ensure we're fixing the "expected bad" input

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+
+    private static LinearLayout addonsDownloadButton;
+
+    static Activity addonsBrowserActivity = null;
```
This leaks a context, let's remove it and try another design

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+@SuppressWarnings("deprecation")
+public class AddonsBrowser extends NavigationDrawerActivity implements DeckDropDownAdapter.SubtitleListener, AddonsAdapter.OnAddonClickListener {
+
+    private static RecyclerView addonsListRecyclerView;
```
I don't think any of these variables should be static. It will probably involve making methods non-static

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+            // reset for new view create
+            addonsNames.clear();
+
+            File directory = new File(String.valueOf(addonsHomeDir));
```
TODO: I don't think you need this

```suggestion
            File directory = addonsHomeDir;
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+
+                // Read package.json from
+                // AnkiDroid/addons/ankidroid-addon-../package/package.json
+                File addonsPackageDir = new File(file, "package");
+                File addonsPackageJson = new File(addonsPackageDir, "package.json");
```
I think this will work

```suggestion
                File addonsPackageJson = new File(file, "package/package.json");
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+                AddonModel addonModel;
+                if (isValidAddonPackage(jsonObject)) {
+                    addonModel = AddonModel.tryParse(jsonObject);
+                    if (addonModel.getType().equals(addonType)) {
```
Extract this logic in the `if` statement into `AddonModel`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+            });
+        }
+
+        hideProgressBarStatic();
```
Move this outside the method and into a `try...finally` on the one caller that doesn't have it already.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+
+
+    public static void deleteDirectory(File dir) {
+        if (!dir.isDirectory()) {
```
This would be useful to extract to a `Util` class, it's almost the same as `BaclupManager.removeDir`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+        for (Map.Entry<String, ?> entry : keys.entrySet()) {
+            Timber.d("map values %s: %s", entry.getKey(), entry.getValue());
+
+            if (entry.getKey().contains("reviewer_addon") && entry.getValue().equals(true)) {
```
Instead of having multiple preferences, a `StringSet` might be the way to go.

We might also want to store this data in `MetaDB`, as it's per-device

david-allison-1 replied:
Invert the if for readability

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,386 @@
+                        content.append("<script>");
+                        content.append("\n");
+
+                        BufferedReader br = new BufferedReader(new FileReader(addonsContentFile));
+                        String line;
+
+                        while ((line = br.readLine()) != null) {
+                            content.append(line);
+                            content.append('\n');
+                        }
+                        br.close();
```
Extract, or use a Util method

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,249 @@
+            JSONObject jsonObject = new JSONObject(strResponse);
+            if (isValidAddonPackage(jsonObject)) {
+
+                listAddonType = jsonObject.optString("addon_type", "");
```
Extract these accessors to the class as well

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,249 @@
+            }
+
+        } catch (JSONException e) {
+            hideProgressBarStatic();
```
Do you want this in the `catch`, or in a `finally`?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,249 @@
+
+        } catch (JSONException e) {
+            hideProgressBarStatic();
+            Timber.e(e.getLocalizedMessage());
```
```suggestion
            Timber.w(e);
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,249 @@
+        JSONObject jsonObject = null;
+        try {
+
+            InputStream is = new FileInputStream(addonsFiles);
```
Does this ned a Try with resources?

infinyte7 replied:
Yes, It shows for file not found exception.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,249 @@
+     */
+    public static void getPackageJson(String npmAddonName, Context context) {
+        showProgressBarStatic();
+        String url = "https://registry.npmjs.org/" + npmAddonName + "/latest";
```
Extract to constant

david-allison-1 reviewed (changes requested):
We're performing a lot of `new File()` on user input. 

This is unsafe in Java: https://stackoverflow.com/a/44496536/13121290

I believe this is the same in Android. Could you test?

----

Code style: https://github.com/ankidroid/Anki-Android/wiki/Code-style#non-public-non-static-field-names-should-start-with-m

infinyte7 commented:
I have remove static implementation. It shows progressbar and toast. But when addons extracted to addons folder, I find difficult to refresh recyclerview with new addons or call `listAddonsFromDir`.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2265,6 +2255,11 @@ private void updateCard(final String newContent) {
         }
 
+        // get all enabled addons' content before so called only once during review time instead of calling each time
+        if (AnkiDroidApp.getSharedPrefs(this).getBoolean("javascript_addons_support", false)) {
```
Move this into the `mCardTemplate` class - it's designed to speed up rendering the template, and the addons can be permanently inserted into the string at the correct position so the string needs less allocations

infinyte7 replied:
I will implement this.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -274,12 +267,13 @@ public void onAddonClick(int position) {
                         SharedPreferences.Editor editor = preferences.edit();
-                        String addonFullName = AddonModel.getAddonFullName(addonModel);
-                        editor.remove(addonFullName).apply();
+                        editor.putStringSet("enabledAddons", enabledAddonSet).apply();
```
Implementers choice - handling the `StringSet/Preferences` might be better off in a class to reduce the code here

infinyte7 replied:
To be implemented.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -299,88 +293,86 @@ public void onAddonClick(int position) {
-        for (Map.Entry<String, ?> entry : keys.entrySet()) {
-            Timber.d("map values %s: %s", entry.getKey(), entry.getValue());
+        for (String s : enabledAddonSet) {
+            if (!s.contains("reviewer_addon")) {
```
Do you need this? Could you have one set for reviewer addons, and one for note editor?

This would also allow you to simplify later code

infinyte7 replied:
I am implementing for separate addons. It will be better to handle separately.

david-allison-1 replied:
OK. I was thinking one preference key per list of addons would make things much more simpler

infinyte7 replied:
Yes, I understood. I am implementing this.
`
Set<String> reviewerEnabledAddonSet = preferences.getStringSet("reviewerAddonKey", new HashSet<String>());
Set<String> noteEditorEnabledAddonSet = preferences.getStringSet("noteEditorEnabledAddonSet", new HashSet<String>());

`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -299,88 +293,86 @@ public void onAddonClick(int position) {
+    public static String readIndexJs(File addonsContentFile) throws IOException {
+        StringBuilder content = new StringBuilder();
+        if (addonsContentFile.exists()) {
+            BufferedReader br = new BufferedReader(new FileReader(addonsContentFile));
```
Use a try with resources, extract the file reading further out

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -299,88 +293,86 @@ public void onAddonClick(int position) {
-                    }
+    public static String readIndexJs(File addonsContentFile) throws IOException {
+        StringBuilder content = new StringBuilder();
+        if (addonsContentFile.exists()) {
```
invert the if

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -25,27 +26,29 @@
-    // Update if api get updated
-    private static String AnkiDroidJsAPI = "0.0.1";
-    private static String AnkiDroidJsAddonKeywords = "ankidroid-js-addon";
+    private String addonType = "reviewer";
```
`static final` since it's constant, you'll need a rename inside Android Studio

```java
    private static final String ADDON_TYPE = "reviewer";
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -25,27 +26,29 @@
-        showProgressBarStatic();
-        String url = "https://registry.npmjs.org/" + npmAddonName + "/latest";
+    public void getPackageJson(String npmAddonName) {
+        showProgressBar();
```
Where has the `hideProgressBar()` gone?

infinyte7 replied:
I am implementing it.

david-allison-1 reviewed (commented):
Much nicer - this is a brief review due to time constraints, will get more structural points on the next round

My main worry is that `StringJoiner` doesn't seem to fix the underlying issue of someone passing in `../`

infinyte7 commented:
For file path, i tried using `../` but it shows error. Here only user supplied value is `addonDir` but it also id for npm package and only possible to change if user want to edit it.
It is not possible to put `../` when downloaded from npmjs as it is unique id. 
```
                StringJoiner joinedPath = new StringJoiner("/")
                        .add(currentAnkiDroidDirectory)
                        .add("addons")
                        .add(addonDir)
                        .add("package")
                        .add("index.js");
```
![1](https://user-images.githubusercontent.com/12841290/113516543-69181980-95ad-11eb-9f53-feef7a531f0c.png)

infinyte7 commented:
It become more simpler than previous and I have noticed performance improved a lot after using `CardTemplate`
Thanks

infinyte7 commented:
Should I move just this one function to AddonsBrowser instead of AddonsNpmUtility ?
```
public void getLatestPackageJson(String npmAddonName) {
        showProgressBar();

        String url = getString(R.string.ankidroid_js_addon_npm_registry, npmAddonName);

        OkHttpClient client = new OkHttpClient();

        Request request = new Request.Builder()
                .url(url)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                Timber.e("js addon %s", e.toString());
            }

            @Override
            public void onResponse(Call call, Response response) {
                try {
                    String strResponse = response.body().string();
                    runOnUiThread(() -> parseJsonData(strResponse, npmAddonName));
                } catch (IOException | NullPointerException e) {
                    Timber.e(e.getLocalizedMessage());
                }
            }
        });

    }
```

Because I need to call `listAddonsFromDir` to update recyclerview after addons download/extracted to addons folder. But that needs to change to static, which to change of many variables to static.
But changing `getLatestPackageJson` to AddonBrowser, it become very easy to call `listAddonsFromDir` also no static call.
I tried some other approach but didn't work.

david-allison-1 commented:
>         new Thread(() -> {
> 
>             runOnUiThread(() -> {
>                 String res = fetchThroughHttp(url);
>                 Timber.d("res:: %s", res);
>             });
> 
>         });

You need to call `thread.start()`

david-allison-1 commented:
> Should I move just this one function to AddonsBrowser instead of AddonsNpmUtility ?
> 
> ```java
> public void getLatestPackageJson(String npmAddonName) {
>         showProgressBar();
> 
>         String url = getString(R.string.ankidroid_js_addon_npm_registry, npmAddonName);
> 
>         OkHttpClient client = new OkHttpClient();
> 
>         Request request = new Request.Builder()
>                 .url(url)
>                 .build();
> 
>         client.newCall(request).enqueue(new Callback() {
>             @Override
>             public void onFailure(Call call, IOException e) {
>                 Timber.e("js addon %s", e.toString());
>             }
> 
>             @Override
>             public void onResponse(Call call, Response response) {
>                 try {
>                         String strResponse = response.body().string();
>                         npmUtility.parseJsonData(strResponse, npmAddonName);
> 
>                         //runOnUiThread(() -> {
>                             //listAddonsFromDir(addonType);
>                         //});
> 
>                 } catch (IOException | NullPointerException e) {
>                     Timber.e(e.getLocalizedMessage());
>                 }
>             }
>         });
> 
>     }
> ```
> 
> Because I need to call `listAddonsFromDir` to update recyclerview after addons download/extracted to addons folder. But that needs to change to static, which to change of many variables to static.
> But changing `getLatestPackageJson` to AddonBrowser, it become very easy to call `listAddonsFromDir` also no static call.
> I tried some other approach but didn't work.
> 
> I tried `fetchThroughHttp` from HttpFetcher but got following errors.

You can pass in a regular method as a callback, you shouldn't need a static method

david-allison-1 commented:
> For file path, i tried using `../` but it shows error on package because from there `../` can be input. Here only user supplied value is `addonDir` and also it is id for npm package and only possible to change if user manually want to edit it after install.
> It is not possible to put `../` when downloaded from npmjs as it is unique id.

Our code shouldn't be trusting external conventions of third parties for security - we should ensure explicitly that the path(s) are rooted where we expect them to be rooted.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -2251,7 +2251,13 @@ private void updateCard(final String newContent) {
+            addons = AddonsBrowser.getEnabledAddonsContent(this);
+        }
+
+        mCardContent = mCardTemplate.render(content, style, cardClass, addons);
```
We can do better - addons is effectively a constant, so we can load it/insert it into `mCardTemplate`

infinyte7 replied:
Should I create `static final String addons` in class? or other method.

david-allison-1 replied:
There's no need for another variable - we can mutate the variables inside the class that we have already:

* One of the variables inside `CardTemplate` will have `::addons::`
* add a `setAddons()` method (or do it in the constructor if this is viable)
* Loop through each variable, and replace `::addons::` with the content provided

Now, we have the same number of concatenations per card, and we've included the addon content

infinyte7 replied:
Thanks, I am implementing it.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,145 @@
+            return false;
+        }
+
+        if (!addonModel.getJsApiVersion().equals(AnkiDroidJsAPI) && jsAddonKeywordsPresent) {
```
Could you add a comment to explain why we have this check

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        // if other strings are non empty
+        if (!addonModel.getName().isEmpty() && !addonModel.getVersion().isEmpty() && !addonModel.getDeveloper().isEmpty() && !addonModel.getHomepage().isEmpty()) {
```
`addonModel.isValid()` perhaps?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    public void updatePrefs(SharedPreferences preferences, String reviewerAddonKey, String addonName, boolean remove) {
+        Set<String> reviewerEnabledAddonSet = preferences.getStringSet(reviewerAddonKey, new HashSet<String>());
+        SharedPreferences.Editor editor = preferences.edit();
+
+        Set<String> newStrSet = new HashSet<String>();
+        newStrSet.addAll(reviewerEnabledAddonSet);
+        newStrSet.add(addonName);
+
+        if (remove) {
+            newStrSet.remove(addonName);
+        }
+
+        editor.putStringSet(reviewerAddonKey, newStrSet).apply();
```
Can't you edit `reviewerEnabledAddonSet` without copying the set?

```suggestion
        Set<String> reviewerEnabledAddonSet = preferences.getStringSet(reviewerAddonKey, new HashSet<String>());

        if (remove) {
            reviewerEnabledAddonSet.remove(addonName);
        } else {
            reviewerEnabledAddonSet.add(addonName);
        }

        preferences.edit().putStringSet(reviewerAddonKey, newStrSet).apply();
```

infinyte7 replied:
I tried but didn't worked so I found solution here https://stackoverflow.com/questions/19949182/android-sharedpreferences-string-set-some-items-are-removed-after-app-restart/19949833
It needs to copy. Also tried above but preferences not working.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,160 @@
+     */
+    public void parseJsonData(String strResponse, String npmAddonName) {
+        try {
+            Timber.d("json::%s", strResponse);
```
```suggestion
            Timber.v("json::%s", strResponse);
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,160 @@
+            if (AddonModel.isValidAddonPackage(jsonObject, ADDON_TYPE)) {
+
+                JSONObject dist = jsonObject.getJSONObject("dist");
+                String tarballUrl = dist.get("tarball").toString();
```
Can this be done?
```suggestion
                String tarballUrl = dist.getString("tarball");
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,145 @@
+            Timber.w(e, e.getLocalizedMessage());
+        }
+
+        if (addonModel == null) {
```
Move this check closer to where `addonModel` is created

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,425 @@
+                            listAddonsFromDir(addonType);
+                        });
+
+                        call.cancel();
+                        response.close();
```
Do you need these if the call if not successful?

infinyte7 replied:
I viewed that in logcat so I implemented above code.
```
WARNING: A connection to https://npmjs.com/ was leaked. Did you forget to close a response body?
```

david-allison-1 reviewed (commented):
@mikehardy @Arthur-Milchior 

Opinions/reviews would be great

Main current question: what happens if the user is offline or if the internet fails, are the errors understandable?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    public void updatePrefs(SharedPreferences preferences, String reviewerAddonKey, String addonName, boolean remove) {
+        Set<String> reviewerEnabledAddonSet = preferences.getStringSet(reviewerAddonKey, new HashSet<String>());
```
Could you add a reference to SO, showing that you can't modify the set

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,145 @@
+
+    public void updatePrefs(SharedPreferences preferences, String reviewerAddonKey, String addonName, boolean remove) {
+        Set<String> reviewerEnabledAddonSet = preferences.getStringSet(reviewerAddonKey, new HashSet<String>());
+        SharedPreferences.Editor editor = preferences.edit();
+
+        Set<String> newStrSet = new HashSet<String>();
+        newStrSet.addAll(reviewerEnabledAddonSet);
+        newStrSet.add(addonName);
+
+        if (remove) {
+            newStrSet.remove(addonName);
+        }
+
+        editor.putStringSet(reviewerAddonKey, newStrSet).apply();
```
Okay, little cleaner:

```suggestion
        Set<String> newStrSet = new HashSet<String>(reviewerEnabledAddonSet);
        
        if (remove) {
            newStrSet.remove(addonName);
        } else {
            newStrSet.add(addonName);
        }

        preferences.edit().putStringSet(reviewerAddonKey, newStrSet).apply();
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,425 @@
+
+                // if string is:  npm i ankidroid-js-addon-progress-bar
+                if (npmAddonName.startsWith("npm i")) {
+                    npmAddonName = npmAddonName.replaceAll("npm i", "");
```
We only want a substring, not replacing all instances

```suggestion
                    npmAddonName = npmAddonName.substring("npm i".length());
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,425 @@
+                String npmAddonName = downloadEditText.getText().toString();
+
+                // if string is:  npm i ankidroid-js-addon-progress-bar
+                if (npmAddonName.startsWith("npm i")) {
```
This might be useful as `npmAddonName = StringUtil.trimPrefix("npm i")`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,425 @@
+
+                JSONObject jsonObject = AddonsNpmUtility.packageJsonReader(addonsPackageJson);
+
+                // TODO
```
flagging todo

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,425 @@
+            return content.toString();
+        }
+
+        FileReader fileReader = new FileReader(addonsContentFile);
+        BufferedReader br = new BufferedReader(fileReader);
+        String line;
+        while ((line = br.readLine()) != null) {
+            content.append(line);
+            content.append('\n');
+        }
+        br.close();
```
This needs a try with resources, or use an existing utility method if one exists

infinyte7 replied:
There exist utility method but `content.append('\n');` not present and it makes js code to not work. So I created above.
I will add it to `try...catch`

david-allison-1 commented on AnkiDroid/src/main/res/layout/addon_install_from_npm.xml:
```diff
@@ -0,0 +1,33 @@
+    <Button
+        android:padding="8dp"
+        android:id="@+id/addon_download_button"
+        android:text="@string/sync_conflict_remote"
```
I'd define another string with a `comment=` for "download"

david-allison-1 commented:
I'd like to step back and see what the others have to say: I think it's in a much much better place now and it'll be easier to discuss big picture things (if any need to be). @mikehardy @Arthur-Milchior

Main thoughts:
* Security - mostly on the `new File`
* Error handling - addon execution - this is a big one, as we'll be taking on a massive support hit if we get this wrong
* Logging - are we logging an appropriate amount to diagnose issues
* Error handling - addon download
* Documentation and simplification of the process - can we improve this? Do we need to show a download screen
* Privacy - I think we're OK here, but it's best to get explicit confirmation before opening NPM

infinyte7 commented:
> > Because I need to call `listAddonsFromDir` to update recyclerview after addons download/extracted to addons folder. But that needs to change to static, which to change of many variables to static.
> > But changing `getLatestPackageJson` to AddonBrowser, it become very easy to call `listAddonsFromDir` also no static call.
> > I tried some other approach but didn't work.
> > I tried `fetchThroughHttp` from HttpFetcher but got following errors.
> 
> You can pass in a regular method as a callback, you shouldn't need a static method

I need help in creating the call back function. I tried but it seems not easy for me.

david-allison-1 commented:
> I need help in creating the call back function. I tried but it seems not easy for me.

What is the method signature that you need? Typically `Runnable` for 0 args, 0 return value

infinyte7 commented:
void with one args

david-allison-1 commented:
> void with one args

`Consumer<T>`, or write an interface with the given method signature.

infinyte7 commented:
Okay, I am implementing it.

infinyte7 commented:
I have implemented using `Runnable` and it give the expected result.
Thanks

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -50,7 +51,7 @@ public AddonsNpmUtility(Context context, Activity activity) {
      * @param npmAddonName addon name, e.g ankidroid-js-addon-progress-bar
      */
-    public void getPackageJson(String npmAddonName) {
+    public void getPackageJson(String npmAddonName, Runnable runnable) {
```
Explain when the runnable is run in the documentation

david-allison-1 commented:
I think the next stage would be to produce an addon that crashes for local testing - I want to reduce the blame for the app crashing/breaking as much as possible, and this will allow us to test mitigation strategies

infinyte7 commented:
Okay, I will implement tests for crashing app. 
Thanks

david-allison-1 commented:
@infinyte7 If you rebase this onto the latest master, then `gradlew lintRelease`/rebuild, the IDE will pick up variable naming issues

infinyte7 commented:
I have implemented the test in `CardTemplateTest.java` and `AddonModelTest.java`. I get the help from `AnalyticsTest.java` for prefs but still getting error. Also for json I viewed existing implemented code. But build failed to pass.

TarekkMA commented:
The issue with the tests is that there are 2 dependencies that are static. `AnkiDroidApp` and `AddonsNpmUtility`, thanks to mockito we can mock static methods with ease using one of the following syntaxes

##### Using Try Resources Syntax
```java
 try (MockedStatic<AnkiDroidApp> utilities = Mockito.mockStatic(AnkiDroidApp.class)) {
    utilities.when(() -> AnkiDroidApp.getSharedPrefs(mMockContext)).thenReturn(mMockSharedPreferences)
}
```
##### Manually closing them
```java
private MockedStatic<AnkiDroidApp> mMockAnkiDroidApp;
....
mMockAnkiDroidApp = Mockito.mockStatic(AnkiDroidApp.class);
....
mMockAnkiDroidApp.when(() -> AnkiDroidApp.getSharedPrefs(mMockContext)).thenReturn(mMockSharedPreferences);
....
mMockAnkiDroidApp.close();
```

And the two stub methods we need will look something like this:
```java
mMockAnkiDroidApp.when(() -> AnkiDroidApp.getSharedPrefs(mMockContext)).thenReturn(mMockSharedPreferences);
mMockAddonsNpmUtility.when(() -> AddonsNpmUtility.getEnabledAddonsContent(mMockContext)).thenReturn("addon");
```

infinyte7 commented:
Thanks @TarekkMA

infinyte7 commented:
@david-allison-1 Now There are two test implemented AddonModel and CardTemplate. What would be next step?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,285 @@
+
+    private SharedPreferences mPreferences;
+
+    private final String mADDON_TYPE = "reviewer";
```
```suggestion
    private static final String ADDON_TYPE = "reviewer";
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+
+public class AddonsNpmUtility {
+
+    private static final String ADDON_TYPE = "reviewer";
```
This seems duplicated - make one public

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+        JSONObject jsonObject = null;
+        try {
+
+            InputStream is = new FileInputStream(addonsFiles);
```
Use a try-with-resources block here, not just a try-catch

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+
+            InputStream is = new FileInputStream(addonsFiles);
+            String stream = Utils.convertStreamToString(is);
+            jsonObject = new JSONObject(stream);
```
You can return here

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+        } catch (FileNotFoundException | JSONException e) {
+            Timber.e(e.getLocalizedMessage());
+        }
+        return jsonObject;
```
Return null if we return early

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+     * @param addonsFiles package.json file in ankidroid-js-addon/package/...
+     * @return jsonObject json object of the addon files content
+     */
+    public static JSONObject packageJsonReader(File addonsFiles) {
```
Mark this as Nullable

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+        // set of enabled addons only
+        Set<String> reviewerEnabledAddonSet = preferences.getStringSet(AddonModel.getReviewerAddonKey(), new HashSet<String>());
+
+        for (String addonDir : reviewerEnabledAddonSet) {
```
Need to define what happens here (ideally with a test) if an addon is defined in preferences, but doesn't exist on the disk

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+        }
+
+        try {
+            FileReader fileReader = new FileReader(addonsContentFile);
```
Needs a few "Try with resources" blocks

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,303 @@
+            FileReader fileReader = new FileReader(addonsContentFile);
+            BufferedReader br = new BufferedReader(fileReader);
+            String line;
+            while ((line = br.readLine()) != null) {
```
There's likely a better method for this

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/AddonModelTest.java:
```diff
@@ -0,0 +1,131 @@
+
+@RunWith(AndroidJUnit4.class)
+public class AddonModelTest extends RobolectricTest {
+    private static final String progrssBarAddonPackageJson = "{\n" +
```
Would make more sense to have this as a file and load it into a string - Java idioms would make diffing this string difficult

david-allison-1 reviewed (commented):
Things that worry me that would be useful in tests

* CardViewer - what happens with an addon throwing JS errors
* What happens if a download fails, or the user is not connected to the internet
* How does a user uninstall an addon
* If an addon is placed on disk in a corrupt form, can the user delete it? 
* What if there's a mismatch between the preferences and what's in the directory?

infinyte7 commented:
> CardViewer - what happens with an addon throwing JS errors

Added a test

> What happens if a download fails, or the user is not connected to the internet

Implemented method for checking if connected to internet

> How does a user uninstall an addon

There is options in info popup to update and delete addon, when clicked on addon

> If an addon is placed on disk in a corrupt form, can the user delete it?

I have tested manually by enabling three addons in dir enable and delete. It removes key from StringSet prefs if addon does not exist on disk.

> What if there's a mismatch between the preferences and what's in the directory?

Here the file downloaded from npmjs so every package have unique name. So, no duplicate in file name. 
`index.js` directly read from `AnkiDroid/addons/some-js-addon/package/index.js` if folder does not exists, then enabled prefs removed for that addon and read operation continue to next addon. 

The structures of addons will be
```
AnkiDroid
   ↳ addons
          ↳ progressbar-addon
                  ↳package
                        ↳ index.js
 
          ↳ card-info-addon
                  ↳ package
                        ↳ index.js
 
          ↳ button-addon
                  ↳ package
                        ↳ index.js
       .....
       .....
   collection.media
    ...
    ...
```

david-allison-1 commented:
Let's get this moving. I don't want to be the only one on this review.

@mikehardy @Arthur-Milchior When you guys have time, could you do an initial scan of the PR? 

@infinyte7 Do you have time now to work on additional feedback, or is it close to exam time for you?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,157 @@
+
+        // js addons have full access to AnkiDroid JS API so needs to check the version
+        // jsAddonKeywords help in distinguish from other npm package
+        if (!addonModel.getJsApiVersion().equals(AnkiDroidJsAPI) || !jsAddonKeywordsPresent) {
```
How is compatibility handled (AnkiDroid too new, or addon too new)? Could this be documented.

Is it possible to set this to a range?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonModel.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    /**
+     * @param preferences SharedPreferences
```
The javadocs don't explain what this method does, especially `reviewerAddonKey`, which seems to be the most important aspect

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,106 @@
+        String reviewerAddonKey = AddonModel.getReviewerAddonKey();
+        Set<String> reviewerEnabledAddonSet = mPreferences.getStringSet(reviewerAddonKey, new HashSet<String>());
+
+        for (String s : reviewerEnabledAddonSet) {
```
Could this be simplified to `if (reviewerEnabledAddonSet.contains(addonModel.getName())`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:
```diff
@@ -0,0 +1,106 @@
+        // turn on switch status else it is off by default
+
+        // store enabled/disabled status as boolean true/false value in SharedPreferences
+        String reviewerAddonKey = AddonModel.getReviewerAddonKey();
```
I feel referring to the constant here would be more readable

infinyte7 commented:
There are three exams in this month. So, I will be able to contribute to this PR from first week in July.

Reviewing big PR takes time.
So, I have decided to re-submit PR in small size. So, that will be easier to review. I will use the same code in that PR.

PR 1
Implementation of downloaing addon from npm.org (.tgz file) with test 

PR 2
Install, update and delete the of addons files (Recyclerview showing list of addons) with test

PR 3
Adding contents of addon data to reviewer with test

The current PR needs some improvement and test case. But this PR is also complete. If this PR is reviewed then that will also be good.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,282 @@
+
+import static com.ichi2.anim.ActivityTransitionAnimation.Direction.END;
+
+@SuppressWarnings("deprecation")
```
Typically it's best to mark the specific methods and add a reference to the GitHub issue if you suppress deprecation

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsBrowser.java:
```diff
@@ -0,0 +1,282 @@
+        getSupportActionBar().setTitle(getString(R.string.reviewer_addons));
+        showBackIcon();
+
+        mNpmUtility = new AddonsNpmUtility(AddonsBrowser.this, this);
```
Typically you can inline this to the variable as long as you're not doing work in the `AddonsNpmUtility` constructor

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,322 @@
+
+public class AddonsNpmUtility {
+
+    public static final String ADDON_TYPE = "reviewer";
```
This will need a more descriptive name `ADDON_TYPE_REVIEWER`

david-allison-1 reviewed (commented):
I've gone through about half again. So much better than it was. 

My main questions: 
* how do we handle JS errors and display to the user that it's an addon problem? Can we improve the UX (and possibly suggest disabling an addon)?
   * Can we attribute an error to a specific addon (or a clash between addons)?
* How are breaking changes handled, from both the AnkiDroid and Addon side?

infinyte7 commented:
> how do we handle JS errors and display to the user that it's an addon problem? Can we improve the UX (and possibly suggest disabling an addon)?

All content of js file read then added to reviewer (Same as adding js code to card template). So, any error in js, that will be in webview. It needs to track the error origin. Like in which file errors occur then notify user that the error occur in the js file of addons and needs to update/delete addons. It can be implemented.

> Can we attribute an error to a specific addon (or a clash between addons)?

The clash between can happened. But from AnkiDroid side it will/may not catch. Because lets say two addons have same variable name then AnkiDroid will not tell there is errors. (This will make AnkiDroid more like js interpreter. So, I think it will be left to addon developer to choose/decide/develop). But if this features needed then I may need to find solution for it.

> How are breaking changes handled, from both the AnkiDroid and Addon side?

The breaking changes handled using version of AnkiDroid JS API. If version greater than or matched to current api version then it will work with that version of AnkiDroid and JS API otherwise addons js content will not be added to reviewer.

david-allison-1 commented:
Thanks, if these are going to be massive changes, I'd much rather have a quick discussion with all the maintainers to make sure that this is going in the right direction. I don't want to be asking for busy work, and I want to see this in

infinyte7 commented:
I think catching error is easier because in webview if errors occur the Android webview's `onConsoleMessage` may tell file and line number of errors and that info will be used to notify users as well as addon developers.

david-allison-1 commented:
Is it possible to move each addon to a file, rather than concatenating the scripts into a `<script>` tag, this would allow rough identification of where a problem occurs?

infinyte7 commented:
yes, that will be better implementation. Instead of reading content of file appending in `<script>::addons::</script>`
This will be better `<script src="../addons/some-addons/package/index.js">`. It will help in identify error more easily.

infinyte7 commented on AnkiDroid/src/main/java/com/ichi2/anki/AddonsNpmUtility.java:
```diff
@@ -0,0 +1,322 @@
+     *                </p>
+     * @return content of index.js file for every enabled addons in script tag.
+     */
+    public static String getEnabledAddonsContent(Context context) {
```
TODO:
Instead read and append to <script>::addons::</script>
Do this `<script src="../addons/some-addons/package/index.js"></script>`

It is easy to catch the bugs and also reduce complexity of this PR.

infinyte7 commented:
The same PR in kotlin already created.
